### DISCUSSION TITLE
`Preview mode` for Nuxt 3, a simple plugin for SSG/Full Static Mode

### DISCUSSION THREAD

**Discussion by ghost** [nuxt/nuxt#18407] [Ideas]
I have figured out how to implement a `Preview mode` for SSG in Nuxt 3.
I wrote a plugin that does the same for NUXT 3 like [this does for Nuxt 2](https://nuxtjs.org/docs/features/live-preview/)

[Demo on stackblitz](https://stackblitz.com/edit/github-cevhpe-kwtyrw?file=README.md,pages%2Findex.vue,plugins%2Fpreview.client.ts)

Basically I wrote this plugin:

```js
// /plugins/preview.client.ts

export default defineNuxtPlugin((nuxtApp) => {
  const route = useRoute();
  const preview = route.query.preview && route.query.preview === 'true';

  if (preview) {
    nuxtApp.hook('page:finish', () => {
      refreshNuxtData();
    });
  }

  return { provide: { preview } };
});

```

And then used it in my app:

```vue
<script lang="ts" setup>
const { data } = await useAsyncData(() => {
  const { $preview } = useNuxtApp();
  if ($preview) {
    return {
      page: {
        title: 'Hello preview world',
        description: 'Additional info',
      },
    };
  }

  return {
    page: {
      title: 'Hello world',
    },
  };
});
</script>

<template>
  <div>
    <ul>
      <li>
        <a :href="'/'">home (/)</a>
      </li>
      <li>
        <a :href="'/?preview=true'">home with preview (/?preview=true)</a>
      </li>
    </ul>
    <p>{{ data }}</p>
  </div>
</template>

```

If you now add `?preview=true` to your page, a different content will be rendered client side.

**What I am not sure about at the moment**: Is there a better hook than `page:finish`? First I’ve used `app:mounted`, but that sometimes fired too early.

**Comment by saktya** [comment]
Its not working if the page data coming from useFetch

**Comment by NathanDubord** [comment]
This has stopped working in Nuxt 3.5.1, @JonathanDoelan any idea off the top of your head? I can try and find some time this weekend to take a look. It's something to do with the .client in the plugin name.

**Comment by formfcw** [comment]
Here is my solution based on the code above.

[Demo on stackblitz](https://stackblitz.com/edit/github-cevhpe-qy4fvo?file=plugins%2Fpreview.ts,pages%2Findex.vue,components%2FPreview.vue)

```ts
// /plugins/preview.ts

export default defineNuxtPlugin(() => {
  const preview = ref(false);

  const previewMiddleware = async (to) => {
    const isPreview = to.query.preview && to.query.preview === 'true';

    // if you don’t need the part with `to.query.token` just set `preview.value = true`
    if (isPreview && to.query.token) {
      preview.value = { token: to.query.token };
      return;
    }

    // If leaving the preview page, refresh the page to exit preview mode
    if (preview.value) {
      window.location = to.fullPath;
      return;
    }
  };

  addRouteMiddleware('preview', previewMiddleware, { global: true });

  return { provide: { preview } };
});
```

**NOTE** that the plugin is not client-only here!


```vue
<!-- /pages/index.vue -->

<script lang="ts" setup>
const { $preview } = useNuxtApp();

// For the preview request we need to specify a different key, otherwise the cached/generated data will be returned by useAsyncData
const keyPrfx = $preview.value ? '-preview' : '';

const { data } = await useAsyncData(`mydata${keyPrfx}`, () => {
  if ($preview.value) {
    return {
      page: {
        title: 'Hello preview world',
        description: 'Additional info',
      },
    };
  }

  return {
    page: {
      title: 'Hello world',
    },
  };
});
</script>

<template>
  <div>
    <ul>
      <li>
        <a :href="'/'">home (/)</a>
      </li>
      <li>
        <a :href="'/?preview=true&token=123'"
          >home with preview (/?preview=true&token=123)</a
        >
      </li>
    </ul>
    <p>{{ data }}</p>
  </div>
</template>
```

**Note** that we need to provide a different key for `useAsyncData` in SSG-mode (`npm run generate` + `npx serve .output/public/`).

Additionally, you can add a preview component …

```vue
<!-- /components/Preview.vue -->

<template>
  <client-only>
    <div v-if="preview">Preview</div>
  </client-only>
</template>

<script setup>
const { $preview: preview } = useNuxtApp();
</script>

<style scoped>
div {
  position: absolute;
  top: 0;
  left: 50%;
  transform: translate(-50%, 0);
  cursor: default;
  background: black;
  color: white;
  padding: 0.3em;
}
div:hover {
  opacity: 0;
}
</style>
```

… that you would implement in app.vue (or in a layout):


```vue
<!-- /app.vue -->

<template>
  <div>
    <preview />
    <NuxtPage></NuxtPage>
  </div>
</template>
```

**Reply by saktya** [reply]
the 'ref' inside plugin causing problem when used in multiple plugin solution, it thrown "A composable that requires access to the Nuxt instance was called outside of a plugin" error

**Comment by henriklangerhansson** [comment]
> I have figured out how to implement a `Preview mode` for SSG in Nuxt 3. I wrote a plugin that does the same for NUXT 3 like [this does for Nuxt 2](https://nuxtjs.org/docs/features/live-preview/)
> 

I have been using the  `enablePreview()` in Nuxt 2 for Storyblok's visual editor to work as expected but struggle to adapt my code to Nuxt 3. Unfortunately your approach does not work for me, in Nuxt 2 it was as simple as:

```
<!-- preview.client.js -->

export default function ({ app, enablePreview }) {
  if (app.$config.isPreview) {
    enablePreview();
  } 
}
```

but adapting the plugin with `refreshNuxtData()` doesn't seem to have any effect on the Storyblok visual editor. Is there any way to reproduce the behavior of `enablePreview()` in Nuxt 3?

**Reply by ghost** [reply]
My initial approach does not work anymore. But check out [this one](https://github.com/nuxt/nuxt/discussions/18407#discussioncomment-5997866)

**Reply by gridsystem** [reply]
> My initial approach does not work anymore. But check out [this one](https://github.com/nuxt/nuxt/discussions/18407#discussioncomment-5997866)

I haven’t tried your original approach but do you know why it doesn’t work anymore? It seems like a logical way of doing things.
